File: problems/1/python/main.py
from fractions import gcd

# all numbers divisible by these are summed
residues = (3, 5)

# max number to check to (exclusive
max_num = 1000

# returns the result from brute force checking
def brute_force():
	# the end sum
	result = 0
	# loop through every number under max, and manually check
	for i in range(0, max_num):
		# make sure it gets reset 
		add_bool = False
		# manually check all in residues
		for j in residues:
			if i % j == 0:
				# if it passes even one, mark it
				add_bool = True
		# if it is divisible by any residue, add it to the result
		if add_bool:
			result += i
	# return the sum
	return result

# returns the result using accelerated summation techniques
def fast_solution():
	# sums multiples of a up to max (not including)
	def mult(a):
		# use this for max sum (the -1 is because it is exclusive)
		st = (max_num - 1) / a
		# standard summation of a linear sequence
		return (a * (st * (st+1))) / 2
	# s(a)+s(b)-s(lcm(a, b))
	# this part counts the multiples of 3, and 5, but then subtracts off the multiples of 3 and 5 (15), so that they are only counted once
	return mult(residues[0]) + mult(residues[1]) - mult((residues[0] * residues[1]) / gcd(residues[0], residues[1]))

print "Sum of numbers that are multiples of %s up to %s is %s (brute force)" % (" or ".join(map(str, residues)), max_num, brute_force()) 
print "Sum of numbers that are multiples of %s up to %s is %s (fast solution)" % (" or ".join(map(str, residues)), max_num, fast_solution()) 

File: problems/1/c/main.c
#include <stdio.h>
#include <stdlib.h>

// Sums multiples up of k up to _max (exclusive)
int sum_mult(int k, int _max) {
	int max = _max - 1;
	return k * (max/k)*(max/k+1) / 2;
}

// returns gcd of a and b
int gcd(a, b) {
	int c = 0;
	// uses euclidean algorithm
	while (b) {
		c = b;
		b = a % b;
		a = c;
	}
	return a;
}
// returns lcm of a and b, the least common multiple of both a and b
int lcm(a, b) {
	return a * b / gcd(a, b);
}

// returns sum of multiples of a or b up to max (exclusive)
int brute_force(int max, int a, int b) {
	//init sum and loop variables
	int i, sum = 0;
	for (i = 0; i < max; ++i) {
		// if i is divisible by a or b, add to the sum
		if (i % a == 0 || i % b == 0) {
			sum += i;
		}
	}
	//return sum
	return sum;
}

//returns sum of multiples of a or b up to max (exclusive) using summing multiples
int fast_solution(int max, int a, int b) {
	//s(a)+s(b)-s(lcm(a, b))
	//This gives multiples of a and b, but subtracts numbers divisible by both, so it isn't counted twice
	return sum_mult(a, max) + sum_mult(b, max) - sum_mult(lcm(a, b), max);
}

int main(int argc, char *argv[]) {
	//initialize defaults
	int max = 1000, a = 3, b = 5, ci = 1;
	//If the user enters more, replace defaults
	if (argc > ci) max = strtol(argv[ci++], NULL, 10);
	if (argc > ci) a = strtol(argv[ci++], NULL, 10);
	if (argc > ci) b = strtol(argv[ci++], NULL, 10);

	//Print out info for both approaches
	printf("Sum of numbers that are multiples of %d or %d up to %d is %d (brute force)\n", a, b, max, brute_force(max, a, b));
	printf("Sum of numbers that are multiples of %d or %d up to %d is %d (fast solution)\n", a, b, max, fast_solution(max, a, b));
	return 0;
}